### DISCUSSION TITLE
VisualTreeHelper.HitTest alternative in WinUI

### DISCUSSION THREAD

**Discussion by gautambjain** [microsoft/microsoft-ui-xaml#7883] [Q&A] [answered]
In WPF we have VisualTreeHelper.HitTest which is used to detect the mouse hover item in different views (list view, tree view).

What is the alternative method in WinUI? Please help.

How to detect the exact item (tree view, list view) under the mouse pointer (mainly used when dragging and dropping)?

**Comment by castorix** [comment, marked as answer]
Like in your other question, [VisualTreeHelper.FindElementsInHostCoordinates](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.visualtreehelper.findelementsinhostcoordinates?view=winrt-22621#windows-ui-xaml-media-visualtreehelper-findelementsinhostcoordinates(windows-foundation-point-windows-ui-xaml-uielement)) can be used

like  :
 
```
  // ptElement is Point
  IEnumerable<UIElement> elementStack = VisualTreeHelper.FindElementsInHostCoordinates(ptElement, (UIElement)sender);
  foreach (UIElement item in elementStack)
  {
     // Code...
  }
```

**Reply by gautambjain** [reply]
Thanks a lot for your answer. I am pretty close. Have been able to find the exact item as per your above answer and [another answer here](https://github.com/microsoft/microsoft-ui-xaml/discussions/7882).

I used GetCursorPos and ScreenToClient.  But however, the Y coordinate is slightly more (about 40 pixels more). If I subtract 40 pixels then I am able to get the element. I believe we should take into consideration the height of the title bar of the window also? Are you aware of this? 

I think FindElementsInHostCoordinates may be expecting coordinates from the start of the root Xaml element. Where as GetCursorPos / ScreenToClient is from the top left corner of the whole app window.

Kindly let me know. Thanks.

**Reply by castorix** [reply]
> I used GetCursorPos and ScreenToClient. But however, the Y coordinate is slightly more (about 40 pixels more). If I subtract 40 pixels then I am able to get the element. I believe we should take into consideration the height of the title bar of the window also? Are you aware of this?
> I think FindElementsInHostCoordinates may be expecting coordinates from the start of the root Xaml element. Where as GetCursorPos / ScreenToClient is from the top left corner of the whole app window.

 ScreenToClient converts to the client coordinates (below the caption)
 I tested with a StackPanel as root and I apparently get the correct Y as the inserted item is always at right node : 

 ```
POINT pt = new POINT();
 GetCursorPos(out pt);
 ScreenToClient(hWnd, ref pt);
 Point ptElement = new Point(pt.x,pt.y);
```

with the declarations and initializations :

  ```
        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Auto)]
        static extern bool GetCursorPos(out POINT lpPoint);

        [StructLayout(LayoutKind.Sequential)]
        public struct POINT
        {
            public int x;
            public int y;

            public POINT(int X, int Y)
            {
                this.x = X;
                this.y = Y;
            }
        }

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Auto)]
        public static extern bool ScreenToClient(IntPtr hWnd, ref POINT lpPoint);

        private IntPtr hWnd = IntPtr.Zero;

        public MainWindow()
        {
            this.InitializeComponent();
            hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);

	        // Test TreeView
            InitializeSampleTreeView();
        }
```

**Reply by gautambjain** [reply]
Thanks a lot again. Your solution is correct. I was at 125% DPI. So I should scale the coordinates accordingly. When I changed the display settings to 100% DPI, it worked correctly.

**Update**: I had to un-scale the coordinates back to 100% and then send to FindElementsInHostCoordinates api.
